detection_rules/generic_loader.py

```diff
@@ -0,0 +1,176 @@
+# Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+# or more contributor license agreements. Licensed under the Elastic License
+# 2.0; you may not use this file except in compliance with the Elastic License
+# 2.0.
+
+"""Load generic toml formatted files for exceptions and actions."""
+import io
+from pathlib import Path
+from typing import Callable, Dict, Iterable, List, Optional, Union
+
+import pytoml
+
+from .action import TOMLAction, TOMLActionContents
+from .exception import TOMLException, TOMLExceptionContents
+from .misc import parse_rules_config
+from .rule_loader import dict_filter
+from .schemas import definitions
+
+
+RULES_CONFIG = parse_rules_config()
+
+GenericCollectionTypes = Union[TOMLAction, TOMLException]
+GenericCollectionContentTypes = Union[TOMLActionContents, TOMLExceptionContents]
+
+
+def metadata_filter(**metadata) -> Callable[[GenericCollectionTypes], bool]:
+    """Get a filter callback based off item metadata"""
+    flt = dict_filter(metadata)
+
+    def callback(item: GenericCollectionTypes) -> bool:
+        target_dict = item.contents.metadata.to_dict()
+        return flt(target_dict)
+
+    return callback
+
+
+class GenericCollection:
+    """Generic collection for action and exception objects."""
+
+    items: list
+    __default = None
+
+    def __init__(self, items: Optional[List[GenericCollectionTypes]] = None):
+        self.id_map: Dict[definitions.UUIDString, GenericCollectionTypes] = {}
+        self.file_map: Dict[Path, GenericCollectionTypes] = {}
+        self.name_map: Dict[definitions.RuleName, GenericCollectionTypes] = {}
+        self.items: List[GenericCollectionTypes] = []
+        self.errors: Dict[Path, Exception] = {}
+        self.frozen = False
+
+        self._toml_load_cache: Dict[Path, dict] = {}
+
+        for items in (items or []):
+            self.add_item(items)
+
+    def __len__(self):
+        """Get the total amount of exceptions in the collection."""
+        return len(self.items)
+
+    def __iter__(self):
+        """Iterate over all items in the collection."""
+        return iter(self.items)
+
+    def __contains__(self, item: GenericCollectionTypes):
+        """Check if an item is in the map by comparing IDs."""
+        return item.id in self.id_map
+
+    def filter(self, cb: Callable[[TOMLException], bool]) -> 'GenericCollection':
+        """Retrieve a filtered collection of items."""
+        filtered_collection = GenericCollection()
+
+        for item in filter(cb, self.items):
+            filtered_collection.add_item(item)
+
+        return filtered_collection
+
+    @staticmethod
+    def deserialize_toml_string(contents: Union[bytes, str]) -> dict:
+        return pytoml.loads(contents)
+
+    def _load_toml_file(self, path: Path) -> dict:
```

Comment:
May want to decouple this idea, but we can most likely replace this with [tomllib](https://docs.python.org/3/library/tomllib.html)

Reply:
tomllib is write only, but they do recommend this for reads: https://pypi.org/project/tomlkit/

Most intriguing is it is _style preserving_!!